prae_macro/src/lib.rs

```diff
@@ -34,9 +34,9 @@ pub fn define(input: TokenStream) -> TokenStream {
 
     let validate_fn = match &guard {
         GuardClosure::Ensure(EnsureClosure(closure)) => quote! {
-            fn validate(v: &Self::Target) -> Option<prae::ValidationError> {
+            fn validate(v: &Self::Target) -> Option<prae::ValidationError::<#ty>> {
                 let f: fn(&Self::Target) -> bool = #closure;
-                if f(v) { None } else { Some(prae::ValidationError) }
+                if f(v) { None } else { Some(prae::ValidationError::<#ty>::new(stringify!(#ident), v.clone()) ) }
```

Comment:
What if the type is not clone? In this case the user will use `Guarded::mutate` instead of `Guarded::try_mutate`, but they both use `validate`

Reply:
Yeah, exactly. The following code works on HEAD but not on this branch:
```rust
#[derive(Debug)]
struct User {
    name: String,
}

prae::define! {
    ValidUser: User
    ensure |u: &User| !u.name.is_empty()
}
```

Reply:
It's either clone, copy or have the error manage reference lifetimes. I would like to avoid having lifetimes as that constraint will leak into the users code.

Reply:
Maybe we can add a nicer message though, instead of a compile error emanating from the proc macro?

Reply:
I guess it's also can't be done with the current code. Introducing `ConstructionError<E>` and `MutationError<E>` should solve the issue, since we'll be able to insert the `value: T` inside `Guarded::new` and `Guarded::try_mutate`

Reply:
Let's just get rid of `value: T` in the current `ValidationError`, so we can merge changes, and I we'll investigate `ConstructionError<E>` and `MutationError<E>`, maybe it'll work :)

Oh, and we should definitely add a test for non-clonable data, so we'll know that our future changes don't break it

Reply:
I don't see a way around cloning, as we have to clone at some point so the error can store the value, or.... you convert the value to a string which requires `T: Debug` or `T: DIsplay`.

Requiring `Clone` on a type is a reasonable expectation given that it's trivial to implement and it's the exception that you see types that are not clone-able.

Reply:
The latest commit removes the value.

Reply:
Making it `T: Debug` and turning it into string is a reasonable workaround! But expecting the value to be always `Clone` is a no go, since the library must be available for as many use cases as possible. Sometimes making your value `!Clone` is a design choice, and we must tolerate this. 

Do you want to add `dbg_value: String` or should I do it?

Reply:
I can do it. I'll update the non_clone test too.

Reply:
`ValidationError::value` is reinstated as a `String`, changes are pushed.